src/librustc_trans/trans/common.rs

```diff
@@ -656,6 +657,14 @@ impl<'blk, 'tcx> BlockS<'blk, 'tcx> {
                                          self.fcx.param_substs,
                                          value)
     }
+
+    pub fn build(&self) -> Builder<'blk, 'tcx> {
+        // Create a fresh builder from the crate context.
+        let builder = self.ccx().builder();
+        // Set the builder's position to this block's end.
+        builder.position_at_end(self.llbb);
```

Comment:
One of the points (other than getting rid of the unnecessary checks) was to get rid of repositioning the builder for every instruction. This should be not necessary (not in MIR, at least). I think it would be better to have something like

```
struct MIRBlock<...> {
     builder: Builder<...>
     bcx: Block<...>
}

// split out the `Builder` trait implemented by the `Builder` struct in trans/builder, and
impl<...> Builder for MIRBlock<...> {
    // has all the functions that produce LLVM insns (likely generated by a macro or something)
}
```

so you could just call

```
bcx.br(self.llblock(target))
```

anywhere in the MIR, instead of doing the `B(bcx).br(...)` or `bcx.builder().br()` dance every time we emit an instruction.